test/api.js

```diff
@@ -1516,6 +1516,141 @@ describe('/v1', function() {
+              code: res.result.code,
+            }
+          });
+        }).then(function(res) {
+          assert.equal(res.statusCode, 200);
+          assertSecurityHeaders(res);
+          assert.equal(res.result.token_type, 'bearer');
+          assert(res.result.access_token);
+          assert(res.result.refresh_token);
+          assert.equal(res.result.access_token.length,
+            config.get('unique.token') * 2);
+          assert.equal(res.result.refresh_token.length,
+            config.get('unique.token') * 2);
+          assert.ok(res.result.scope.includes('oof baz'));
+          assert.ok(res.result.scope.includes('bar foo'));
+          assert.equal(res.result.auth_at, AUTH_AT);
+        });
+      });
+
+      it('incremental authorization with existing_grant', function() {
+        mockAssertion().reply(200, VERIFY_GOOD);
+        let firstRefreshToken;
+        return Server.api.post({
+          url: '/authorization',
+          payload: authParams({
+            scope: 'bobo bibi',
+            access_type: 'offline'
+          })
+        }).then(function(res) {
+          assert.equal(res.statusCode, 200);
+          assertSecurityHeaders(res);
+          return Server.api.post({
+            url: '/token',
+            payload: {
+              client_id: clientId,
+              client_secret: secret,
+              code: res.result.code,
+            }
+          });
+        }).then(function(res) {
+          assert.equal(res.statusCode, 200);
+          assertSecurityHeaders(res);
+          assert.equal(res.result.token_type, 'bearer');
+          assert(res.result.access_token);
+          firstRefreshToken = res.result.refresh_token;
+          assert(res.result.refresh_token);
+          assert.equal(res.result.access_token.length,
+            config.get('unique.token') * 2);
+          assert.equal(res.result.refresh_token.length,
+            config.get('unique.token') * 2);
+          assert.equal(res.result.scope, 'bobo bibi');
+          assert.equal(res.result.auth_at, AUTH_AT);
+          mockAssertion().reply(200, VERIFY_GOOD);
+          return Server.api.post({
+            url: '/authorization',
+            payload: authParams({
+              scope: 'bubu baba',
+              access_type: 'offline'
+            })
+          });
+        }).then(function(res) {
+          assert.equal(res.statusCode, 200);
+          assertSecurityHeaders(res);
+          return Server.api.post({
+            url: '/token',
+            payload: {
+              client_id: clientId,
+              client_secret: secret,
+              code: res.result.code,
+              existing_grant: firstRefreshToken,
+            }
+          });
+        }).then(function(res) {
+          assert.equal(res.statusCode, 200);
+          assertSecurityHeaders(res);
+          assert.equal(res.result.token_type, 'bearer');
+          assert(res.result.access_token);
+          assert(res.result.refresh_token);
```

Comment:
IIUC, this grants a *new* refresh token that has the scopes of the old one, plus the newly-granted scopes, is that correct?  If so, we should probably explicitly test that it's a new token.

To be honest though, I wonder if it makes sense to re-use the existing refresh token in this case.  Or perhaps to immediately revoke the old refresh token when returning the new one?  We'll have to think about what the right behaviour is from the client side here...